### PR TITLE
Allow positional arguments for group by, order by, limit by

### LINKED ISSUE #2592: Feature request: GROUP BY column number support
Feature request:
Support for GROUP BY 1,2 in queries

Example query:
SELECT col1, col2
FROM table1
GROUP BY 1,2

This would be equivalent to current ClickHouse query
SELECT col1, col2
FROM table1
GROUP BY col1, col2

### PR DESCRIPTION
**Pull request by kssenii**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow positional arguments under setting `enable_positional_arguments`. Closes https://github.com/ClickHouse/ClickHouse/issues/2592.

### CODE REVIEW

File: src/Interpreters/TreeOptimizer.cpp [hunk 1]
```diff
@@ -164,7 +168,7 @@ void optimizeGroupBy(ASTSelectQuery * select_query, const NameSet & source_colum
                     std::back_inserter(group_exprs), is_literal
             );
         }
-        else if (is_literal(group_exprs[i]))
+        else if (is_literal(group_exprs[i]) && !settings.enable_positional_arguments)
```
Comment (Avogar) [inline src/Interpreters/TreeOptimizer.cpp:171]: I think when `enable_positional_arguments=true` we can also remove literals that cannot be indexes.

**Comment by Avogar** [comment]
I think there are some small problems.
```sql
avogar-dev.sas.yp-c.yandex.net :) create table test(x1 Int, x2 Int, x3 Int) engine=Memory()

CREATE TABLE test
(
    `x1` Int,
    `x2` Int,
    `x3` Int
)
ENGINE = Memory

Query id: 21f966b1-65fa-4e29-bb10-cb66a40bb100

Ok.

0 rows in set. Elapsed: 0.023 sec.

avogar-dev.sas.yp-c.yandex.net :) insert into test values (1, 10, 100), (10, 1, 10), (100, 100, 1)

INSERT INTO test VALUES

Query id: af086b77-1dd0-4050-9d03-0d27eff98f98

Ok.

2 rows in set. Elapsed: 0.001 sec.
```

1) 
```sql
avogar-dev.sas.yp-c.yandex.net :) select max(x1) from test order by x1

SELECT max(x1)
FROM test
ORDER BY x1 ASC

Query id: 07b7d229-6ac3-4dbf-9b82-3b3647aca506


0 rows in set. Elapsed: 0.002 sec. 

Received exception from server (version 21.9.1):
Code: 215. DB::Exception: Received from localhost:9000. DB::Exception: Column `x1` is not under aggregate function and not in GROUP BY: While processing x1 ASC. (NOT_AN_AGGREGATE)

avogar-dev.sas.yp-c.yandex.net :) select max(x1) from test order by 1

SELECT max(x1)
FROM test
ORDER BY 1 ASC

Query id: fcaf420e-8c39-4e92-abc9-3350d218ccfc


0 rows in set. Elapsed: 0.002 sec. 

Received exception from server (version 21.9.1):
Code: 47. DB::Exception: Received from localhost:9000. DB::Exception: Unknown column: x1, there are only columns max(x1), 1. (UNKNOWN_IDENTIFIER)
```
Exception messages are differ, but maybe it's not important.

2) 
```sql
avogar-dev.sas.yp-c.yandex.net :) select max(x1), max(x2), max(x3) from test group by x1

SELECT
    max(x1),
    max(x2),
    max(x3)
FROM test
GROUP BY x1

Query id: fa5cee5d-91fe-4187-bbe1-ce44a2ac85d3

┌──x1─┬─max(x2)─┬─max(x3)─┐
│   1 │      10 │     100 │
│ 100 │     100 │       1 │
│  10 │       1 │      10 │
└─────┴─────────┴─────────┘

3 rows in set. Elapsed: 0.002 sec. 

avogar-dev.sas.yp-c.yandex.net :) select max(x1), max(x2), max(x3) from test group by 1

SELECT
    max(x1),
    max(x2),
    max(x3)
FROM test
GROUP BY 1

Query id: 857830c3-7b0b-4608-83bd-7c3699df54ab

┌─max(x1)─┬─max(x2)─┬─max(x3)─┐
│       1 │      10 │     100 │
│     100 │     100 │       1 │
│      10 │       1 │      10 │
└─────────┴─────────┴─────────┘

3 rows in set. Elapsed: 0.002 sec.
```
The outputs look a bit differ.

3)
```sql
avogar-dev.sas.yp-c.yandex.net :) select x3, x2, x1 from test order by 1

SELECT
    x3,
    x2,
    x1
FROM test
ORDER BY 1 ASC

Query id: 1721f158-2935-4b3f-bf99-7c52c2d4c133

┌──x3─┬──x2─┬──x1─┐
│ 100 │  10 │   1 │
│  10 │   1 │  10 │
│   1 │ 100 │ 100 │
└─────┴─────┴─────┘

3 rows in set. Elapsed: 0.002 sec. 

avogar-dev.sas.yp-c.yandex.net :) select x3, x2, x1 from test order by x3

SELECT
    x3,
    x2,
    x1
FROM test
ORDER BY x3 ASC

Query id: 8916cb21-b43e-4605-a916-92e95eb64f6e

┌──x3─┬──x2─┬──x1─┐
│   1 │ 100 │ 100 │
│  10 │   1 │  10 │
│ 100 │  10 │   1 │
└─────┴─────┴─────┘

3 rows in set. Elapsed: 0.002 sec. 
```
If I understand correctly, these queries should be the same, because we should use relative index of column from select list.
```sql
avogar-dev.sas.yp-c.yandex.net :) select x3, x2 from test order by 1

SELECT
    x3,
    x2
FROM test
ORDER BY 1 ASC

Query id: dab55d95-6172-48f7-9d45-6cc8328d9b2f


0 rows in set. Elapsed: 0.001 sec. 

Received exception from server (version 21.9.1):
Code: 47. DB::Exception: Received from localhost:9000. DB::Exception: Unknown column: x1, there are only columns x2, x3, 1. (UNKNOWN_IDENTIFIER)
```
I would expect that this query will be the same as ` select x3, x2 from test order by x3`
Or we should use the column index from the table?

**Comment by Avogar** [comment]
As I can see, in SQL we shoud use relative index except case when we use `select * `. You can try this query here https://sqliteonline.com/:
```sql
CREATE TABLE test (x1 INTEGER, x2 INTEGER);
INSERT INTO test VALUES (1, 2), (2, 1);
select x2, x1 from test order by 1;
```

File: src/Interpreters/TreeOptimizer.cpp [hunk 2]
```diff
@@ -166,7 +170,22 @@ void optimizeGroupBy(ASTSelectQuery * select_query, const NameSet & source_colum
         }
         else if (is_literal(group_exprs[i]))
         {
-            remove_expr_at_index(i);
+            bool keep_position = false;
+            if (settings.enable_positional_arguments)
+            {
+                const auto & value = group_exprs[i]->as <ASTLiteral>()->value;
```
Comment (Avogar) [inline src/Interpreters/TreeOptimizer.cpp:176]:
```suggestion
                const auto & value = group_exprs[i]->as<ASTLiteral>()->value;
```

**Review by Avogar — APPROVED** [review]

**Comment by Avogar** [comment]
@kssenii Let's fix build check and merge it

**Comment by sevirov** [comment]
Internal documentation ticket: DOCSUP-13583

**Comment by dinhthengoc2021998** [comment]
I've updated Clickhouse to the latest version : 21.10.2.15
I've 2 queries like that:
- First query work correctly : 
> set enable_positional_arguments=1;
> select REPORTED_TIME as DAYS, count(*)
> from oneoss_aggregations.DATA_NETWORK_LOCATION_SERVICE_TECHNOLOGY
> group by 1;

- But the second query doesn't work well : 

> set enable_positional_arguments=1;
> select toString(toStartOfDay(REPORTED_TIME)) as DAYS, count(*)
> from oneoss_aggregations.DATA_NETWORK_LOCATION_SERVICE_TECHNOLOGY
> group by 1;

and i got this error: 
>Code: 43. DB::Exception: Illegal value for positional argument in GROUP BY. (ILLEGAL_TYPE_OF_ARGUMENT) (version 21.10.2.15 (official build))

It's seem like if i only use the raw column, this'll be fine. But if i need to use some transform for the raw column, i get an error like above.
Can you explain how this happened and give me some solutions to handle this? 
Thank you!

**Comment by dinhthengoc2021998** [comment]
> @kssenii Actually I used mysql interface to connect Tableau Server to Clickhouse.
Connection is OK until now, but when i try to create a calculated column in Tableau Server- it's like apply some transform functions - and get the error as i described above. 
Meanwhile, if i connect to standard MYSQL Database, i can create and use calculated column without any error.
I think it's will effect on the way i use Tableau Server to connect and use Clickhouse through MYSQL Interface.
Lastly,thank for your answer.

**Comment by dinhthengoc2021998** [comment]
@UnamedRus 
- I checked and got the answer that ODBC is only available on Tableau Desktop, not on Tableau Online or Tableau Server. Then i can publish that datasource to Tableau Online or Tableau Server, it's work. 
 - But i want to find the way how connect directly from Tableau Server to Clickhouse and found that the way to connect Clickhouse through MYSQL Interface. Connection and create Chart with RawColumn is work well until i try to create calculated column. I found that it's because Tableau will send the query which use Position_Arguments in GROUPBY and Clickhouse doesn't understand correctly

**Pull request merged**